README.md
# ELS - Enterprise Linux/EuroLinux Sources - rwho
 
## Usage:
  Checkout branch or tag. Use `make sources` to recreate  SOURCE0 archive.


src/ChangeLog
22-Jul-2000:
	Add some minor additional paranoia.
	Changes from Olaf Kirch to run chrooted to /var/spool/rwho and
	  then optionally drop root privileges.

3-Jun-2000:
	Fix sizes of types in enclosed protocols/rwhod.h.

14-Dec-1999:
	netkit-rwho-0.16 is released.

1-Aug-1999:
	Complete y2k and y2038 audit. Clean up y2038 issues.
	Clean up buffer problems in rwho (apparently unexploitable, 
	  fortunately.)

31-Jul-1999:
	Redid makefiles/config stuff for new confgen version.

23-Sep-1997:
	Added ruptime.
	Fixed a suspicious strncpy in rwhod.

12-Jun-1997:
	netkit-rwho-0.10 released.

08-Jun-1997:
	More adjustments for glibc.

19-May-1997:
	glibc fixes from Red Hat.

13-May-1997:
	Patch from Debian to support selective use of interfaces in 
	rwhod. (via Peter Tobias, tobias@et-inf.fho-emden.de)

05-Apr-1997:
	Added configure script to generate MCONFIG.
	Ensure null termination of certain input data. (Peter Tobias,
	  tobias@et-inf.fho-emden.de)
	Better utmp handling in rwhod.

08-Mar-1997: 
	Split from full NetKit package. 
	Generated this change log from NetKit's.

29-Dec-1996
	NetKit-0.09 released.
	Assorted alpha/glibc patches. (Erik Troan, ewt@redhat.com)
	Assorted bug fixes from Debian. (Peter Tobias, 
	  tobias@et-inf.fho-emden.de)
	Hardened programs against DNS h_length spoofing attacks.
	Use inet_aton() everywhere instead of inet_addr().

22-Aug-1996
	NetKit-B-0.08 released.
	(almost) everything now compiles with lots of warnings turned on.
	Fixed rwhod not to need -lbsd.
	Fixed possible buffer overrun hole in rwhod.

25-Jul-1996
	NetKit-B-0.07A released.

23-Jul-1996
	NetKit-B-0.07 released.
	Integrated a collection of patches that had been lurking on the net,
	  including the 256-ptys support for telnetd and passive mode ftp.
	Major security fixes, including to fingerd, lpr, rlogin, rsh, talkd, 
	  and telnetd. Do *not* use the sliplogin from earlier versions of this
	  package, either.
	Much of the code builds without libbsd.a or bsd includes.
	Massive code cleanup. Almost everything compiles clean with gcc
	  -Wall now. rusers and rusersd do not; patches to rpcgen to fix
	  this would be appreciated if anyone feels like it.
	New maintainer:  David A. Holland, dholland@hcs.harvard.edu

date not known
	NetKit-B-0.06 released.

date not known
	NetKit-B-0.05 released.

date not known
	NetKit-B-0.04 released.

date not known
	NetKit-B-0.03 released.
	Changed rusers, rpc.ruserd, rwall, rpc.rwalld, rwho, rwhod to not
	  require to copy ./include/*.h files into /usr/include



src/rwho/rwho.c
/*
 * Copyright (c) 1983 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

char copyright[] =
  "@(#) Copyright (c) 1983 The Regents of the University of California.\n"
  "All rights reserved.\n";

/*
 * From: @(#)rwho.c	5.5 (Berkeley) 6/1/90
 */
char rcsid[] = "$Id: rwho.c,v 1.7 1999/08/01 20:44:18 dholland Exp $";

#include <sys/param.h>
#include <sys/dir.h>
#include <sys/file.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdio.h>
#include <time.h>
#include <protocols/rwhod.h>

#include "../version.h"

static int utmpcmp(const void *, const void *);
#define	NUSERS	1000

struct myutmp {
	char	myhost[MAXHOSTNAMELEN];
	int	myidle;
	char	myline[PATH_MAX];
	char	myname[16];
	time_t	mytime;
};

#define	WHDRSIZE	((int)(sizeof (wd) - sizeof (wd.wd_we)))

/* 
 * this function should be shared with ruptime.
 */
static int down(time_t rcv, time_t now) {
	return (now - rcv > 11 * 60);
}

int 
main(int argc, char *argv[])
{
	static struct myutmp myutmp[NUSERS];
	int nusers = 0;

	int ch;
	struct direct *dp;
	int cc, width;
	struct whod wd;
	struct whod *w = &wd;
	struct whoent *we;
	struct myutmp *mp;
	int f, n, i;
	DIR *dirp;
	time_t now, myrecvtime, timecorrection;
	int aflg = 0;
	size_t size;

	while ((ch = getopt(argc, argv, "a")) != EOF) {
		switch((char)ch) {
		case 'a':
			aflg = 1;
			break;
		case '?':
		default:
			fprintf(stderr, "usage: rwho [-a]\n");
			exit(1);
		}
	}
	if (chdir(_PATH_RWHODIR) || (dirp = opendir(".")) == NULL) {
		perror(_PATH_RWHODIR);
		exit(1);
	}
	mp = myutmp;
	time(&now);
	while ((dp = readdir(dirp))!=NULL) {
		if (dp->d_ino == 0 || strncmp(dp->d_name, "whod.", 5)) {
			continue;
		}
		f = open(dp->d_name, O_RDONLY);
		if (f < 0) {
			continue;
		}
		cc = read(f, (char *)&wd, sizeof (struct whod));
		if (cc < WHDRSIZE) {
			(void) close(f);
			continue;
		}

		myrecvtime = w->wd_recvtime;
		/*
		 * After 2038 the rwho protocol, which sends int32s, will
		 * start to lose. However, by then, "now" will be an int64.
		 * So assume that if "now" is WAY bigger than myrecvtime, 
		 * the latter needs to be incremented by 2^32. 0x40000000 is
		 * 34 years, so it's _fairly_ safe to assume we won't have
		 * any packets that old in /var/spool/rwho.
		 * Increment the correction by values < 2^31 so it works 
		 * if time_t is still an int32. Save the correction for
		 * use on the login times.
		 */
		timecorrection = 0;
		while (now - myrecvtime + timecorrection > 0x40000000) {
			timecorrection += 0x70000000;
			timecorrection += 0x70000000;
			timecorrection += 0x20000000;
		}

		if (down(myrecvtime, now)) {
			(void) close(f);
			continue;
		}

		cc -= WHDRSIZE;
		we = w->wd_we;
		for (n = cc / sizeof (struct whoent); n > 0; n--) {
			if (aflg == 0 && we->we_idle >= 60*60) {
				we++;
				continue;
			}
			if (nusers >= NUSERS) {
				printf("too many users\n");
				exit(1);
			}

			/*
			 * Copy hostname. rwhod is supposed to make sure
			 * wd_hostname is null-terminated, but be defensive.
			 */
			size = sizeof(w->wd_hostname);
			if (size > sizeof(mp->myhost)) {
				size = sizeof(mp->myhost);
			}
			strncpy(mp->myhost, w->wd_hostname, size);
			mp->myhost[size-1] = 0;

			mp->myidle = we->we_idle;

			/*
			 * rwhod does not guarantee null-termination of
			 * the strings in we_utmp. 
			 */
			assert(sizeof(mp->myline) > 
			       sizeof(we->we_utmp.out_line));
			assert(sizeof(mp->myname) > 
			       sizeof(we->we_utmp.out_name));
			strncpy(mp->myline, we->we_utmp.out_line, 
				sizeof(we->we_utmp.out_line));
			mp->myline[sizeof(we->we_utmp.out_line)] = 0;
			strncpy(mp->myname, we->we_utmp.out_name,
				sizeof(we->we_utmp.out_name));
			mp->myname[sizeof(we->we_utmp.out_name)] = 0;
			mp->mytime = we->we_utmp.out_time + timecorrection;

			nusers++; 
			we++;
			mp++;
		}
		(void) close(f);
	}
	qsort(myutmp, nusers, sizeof(struct myutmp), utmpcmp);
	mp = myutmp;
	width = 0;
	for (i = 0; i < nusers; i++) {
		int j = strlen(mp->myhost) + 1 + strlen(mp->myline);
		if (j > width)
			width = j;
		mp++;
	}
	mp = myutmp;
	for (i = 0; i < nusers; i++) {
		char buf[BUFSIZ];
		snprintf(buf, sizeof(buf), "%s:%s", mp->myhost, mp->myline);
		printf("%-8.8s %-*s %.12s",
		       mp->myname,
		       width,
		       buf,
		       ctime(&mp->mytime)+4);
		mp->myidle /= 60;
		if (mp->myidle) {
			if (aflg) {
				if (mp->myidle >= 100*60)
					mp->myidle = 100*60 - 1;
				if (mp->myidle >= 60)
					printf(" %2d", mp->myidle / 60);
				else
					printf("   ");
			} else
				printf(" ");
			printf(":%02d", mp->myidle % 60);
		}
		printf("\n");
		mp++;
	}
	exit(0);
}

static int
utmpcmp(const void *v1, const void *v2)
{
	const struct myutmp *u1 = (const struct myutmp *)v1;
	const struct myutmp *u2 = (const struct myutmp *)v2;
	int rc;

	rc = strcmp(u1->myname, u2->myname);
	if (rc)
		return (rc);
	rc = strcmp(u1->myhost, u2->myhost);
	if (rc)
		return (rc);
	return (strcmp(u1->myline, u2->myline));
}


src/ruptime/ruptime.c
/*
 * Copyright (c) 1983 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	From: @(#)ruptime.c	5.8 (Berkeley) 7/21/90
 *	From: OpenBSD: ruptime.c,v 1.5 1997/07/06 02:28:55 millert Exp
 */

char copyright[] =
  "@(#) Copyright (c) 1983 The Regents of the University of California.\n"
  "All rights reserved.\n";

char ruptime_rcsid[] = 
  "$Id: ruptime.c,v 1.5 1999/12/12 15:33:39 dholland Exp $";

#include "../version.h"

#include <sys/param.h>
#include <sys/file.h>
#include <dirent.h>
#include <protocols/rwhod.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <time.h>

struct hs {
	char	hs_hostname[MAXHOSTNAMELEN];
	time_t	hs_sendtime;
	time_t	hs_recvtime;
	time_t	hs_boottime;
	int	hs_loadav[3];

	int	hs_nusers;
};

#define	ISDOWN(h)	(now - (h)->hs_recvtime > 11 * 60)
#define	WHDRSIZE	((int)(sizeof (awhod) - sizeof (awhod.wd_we)))

static size_t nhosts, hspace = 20;
static struct hs *hs;
static struct whod awhod;
static time_t now;
static int rflg = 1;

static int hscmp(const void *, const void *);
static int ucmp(const void *, const void *);
static int lcmp(const void *, const void *);
static int tcmp(const void *, const void *);

static const char *interval(time_t tval, const char *updown);
static void morehosts(void);

int
main(int argc, char **argv)
{
	register struct hs *hsp;
	register struct whod *wd;
	register struct whoent *we;
	register DIR *dirp;
	struct dirent *dp;
	int aflg, cc, ch, f, maxloadav;
	unsigned i;
	char buf[sizeof(struct whod)];
	int (*cmp)(const void *, const void *) = hscmp;
	time_t correction;

	aflg = 0;
	while ((ch = getopt(argc, argv, "alrut")) != -1)
		switch((char)ch) {
		case 'a':
			aflg = 1;
			break;
		case 'l':
			cmp = lcmp;
			break;
		case 'r':
			rflg = -1;
			break;
		case 't':
			cmp = tcmp;
			break;
		case 'u':
			cmp = ucmp;
			break;
		default: 
			(void)fprintf(stderr, "usage: ruptime [-alrut]\n");
			exit(1);
		}

	if (chdir(_PATH_RWHODIR) || (dirp = opendir(".")) == NULL) {
		(void)fprintf(stderr, "ruptime: %s: %s.\n",
		    _PATH_RWHODIR, strerror(errno));
		exit(1);
	}
	morehosts();
	hsp = hs;
	maxloadav = -1;
	while ((dp = readdir(dirp))) {
		if (dp->d_ino == 0 || strncmp(dp->d_name, "whod.", 5))
			continue;
		if ((f = open(dp->d_name, O_RDONLY, 0)) < 0) {
			(void)fprintf(stderr, "ruptime: %s: %s\n",
			    dp->d_name, strerror(errno));
			continue;
		}
		cc = read(f, buf, sizeof(struct whod));
		(void)close(f);
		if (cc < WHDRSIZE)
			continue;
		if (nhosts == hspace) {
			morehosts();
			hsp = hs + nhosts;
		}
		wd = (struct whod *)buf;
		snprintf(hsp->hs_hostname, sizeof(hsp->hs_hostname),
			 "%s", wd->wd_hostname);
		/*
		 * y2038 issue. rwhod protocol sends int32 times;
		 * assume they're not more than a few years old, and
		 * correct. Add in values < 0x80000000 so it works if 
		 * time_t's 32-bit.
		 */
		correction = 0;
		while (now - wd->wd_sendtime + correction > 0x40000000) {
		    correction += 0x70000000;
		    correction += 0x70000000;
		    correction += 0x20000000;
		}
		hsp->hs_sendtime = wd->wd_sendtime + correction;
		hsp->hs_recvtime = wd->wd_recvtime + correction;
		hsp->hs_boottime = wd->wd_boottime + correction;
		hsp->hs_loadav[0] = wd->wd_loadav[0];
		hsp->hs_loadav[1] = wd->wd_loadav[1];
		hsp->hs_loadav[2] = wd->wd_loadav[2];

		hsp->hs_nusers = 0;
		for (i = 0; i < 2; i++)
			if (wd->wd_loadav[i] > maxloadav)
				maxloadav = wd->wd_loadav[i];
		we = (struct whoent *)(buf+cc);
		while (--we >= wd->wd_we)
			if (aflg || we->we_idle < 3600)
				hsp->hs_nusers++;
		nhosts++;
		hsp++;
	}
	if (!nhosts) {
		(void)printf("ruptime: no hosts in %s.\n", _PATH_RWHODIR);
		exit(1);
	}
	(void)time(&now);
	qsort(hs, nhosts, sizeof(hs[0]), cmp);
	for (i = 0; i < nhosts; i++) {
		hsp = &hs[i];
		if (ISDOWN(hsp)) {
			(void)printf("%-12.12s%s\n", hsp->hs_hostname,
			    interval(now - hsp->hs_recvtime, "down"));
			continue;
		}
		(void)printf(
		    "%-12.12s%s,  %4d user%s  load %*.2f, %*.2f, %*.2f\n",
		    hsp->hs_hostname,
		    interval((time_t)hsp->hs_sendtime -
			(time_t)hsp->hs_boottime, "  up"),
		    hsp->hs_nusers,
		    hsp->hs_nusers == 1 ? ", " : "s,",
		    maxloadav >= 1000 ? 5 : 4,
			hsp->hs_loadav[0] / 100.0,
		    maxloadav >= 1000 ? 5 : 4,
		        hsp->hs_loadav[1] / 100.0,
		    maxloadav >= 1000 ? 5 : 4,
		        hsp->hs_loadav[2] / 100.0);
	}
	exit(0);
}

static
const char *
interval(time_t tval, const char *updown)
{
	static char resbuf[32];
	int days, hours, minutes;

	if (tval < 0 || tval > 999*24*60*60) {
		(void)snprintf(resbuf, sizeof(resbuf), "%s     ??:??", updown);
		return(resbuf);
	}
	minutes = (tval + 59) / 60;		/* round to minutes */
	hours = minutes / 60; minutes %= 60;
	days = hours / 24; hours %= 24;
	if (days)
		(void)snprintf(resbuf, sizeof(resbuf), "%s %3d+%02d:%02d",
		    updown, days, hours, minutes);
	else
		(void)snprintf(resbuf, sizeof(resbuf), "%s     %2d:%02d",
		    updown, hours, minutes);
	return(resbuf);
}

/* alphabetical comparison */
static 
int
hscmp(const void *a1, const void *a2)
{
	const struct hs *h1 = a1, *h2 = a2;

	return(rflg * strcmp(h1->hs_hostname, h2->hs_hostname));
}

/* load average comparison */
static 
int
lcmp(const void *a1, const void *a2)
{
	const struct hs *h1 = a1, *h2 = a2;

	if (ISDOWN(h1))
		if (ISDOWN(h2))
			return(tcmp(a1, a2));
		else
			return(rflg);
	else if (ISDOWN(h2))
		return(-rflg);
	else
		return(rflg * (h2->hs_loadav[0] - h1->hs_loadav[0]));
}

/* number of users comparison */
static
int
ucmp(const void *a1, const void *a2)
{
	const struct hs *h1 = a1, *h2 = a2;

	if (ISDOWN(h1))
		if (ISDOWN(h2))
			return(tcmp(a1, a2));
		else
			return(rflg);
	else if (ISDOWN(h2))
		return(-rflg);
	else
		return(rflg * (h2->hs_nusers - h1->hs_nusers));
}

/* uptime comparison */
static
int
tcmp(const void *a1, const void *a2)
{
	const struct hs *h1 = a1, *h2 = a2;

	return(rflg * (
		(ISDOWN(h2) ? h2->hs_recvtime - now
			  : h2->hs_sendtime - h2->hs_boottime)
		-
		(ISDOWN(h1) ? h1->hs_recvtime - now
			  : h1->hs_sendtime - h1->hs_boottime)
	));
}

static
void
morehosts(void)
{
	hs = realloc((char *)hs, (hspace *= 2) * sizeof(*hs));
	if (hs == NULL) {
		(void)fprintf(stderr, "ruptime: %s.\n", strerror(ENOMEM));
		exit(1);
	}
}


src/include/protocols/rwhod.h
/*
 * Copyright (c) 1983 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	from: @(#)rwhod.h	5.6 (Berkeley) 4/3/91
 *	$Id: rwhod.h,v 1.3 2000/06/03 16:43:06 dholland Exp $
 */

#ifndef _RWHOD_H_
#define	_RWHOD_H_

/*
 * rwho protocol packet format.
 */
struct	outmp {
	char	out_line[8];		/* tty name */
	char	out_name[8];		/* user id */
	int32_t	out_time;		/* time on */
};

struct	whod {
	char	wd_vers;		/* protocol version # */
	char	wd_type;		/* packet type, see below */
	char	wd_pad[2];
	int32_t	wd_sendtime;		/* time stamp by sender */
	int32_t	wd_recvtime;		/* time stamp applied by receiver */
	char	wd_hostname[32];	/* hosts's name */
	int32_t	wd_loadav[3];		/* load average as in uptime */
	int32_t	wd_boottime;		/* time system booted */
	struct	whoent {
		struct	outmp we_utmp;	/* active tty info */
		int32_t	we_idle;	/* tty idle time */
	} wd_we[1024 / sizeof (struct whoent)];
};

#define	WHODVERSION	1
#define	WHODTYPE_STATUS	1		/* host status */

#define	_PATH_RWHODIR	"/var/spool/rwho"

#endif /* !_RWHOD_H_ */


src/rwhod/daemon.c
/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * From: @(#)daemon.c	8.1 (Berkeley) 6/4/93
 *       NetBSD: daemon.c,v 1.4 1995/02/25 13:41:16 cgd Exp
 */
char daemon_rcsid[] = 
  "$Id: daemon.c,v 1.2 1996/08/15 07:22:28 dholland Exp $";

#include <fcntl.h>
#include <paths.h>
#include <unistd.h>
#include "daemon.h"

int daemon(int nochdir, int noclose)
{
	int fd;

	switch (fork()) {
	  case -1:
		return -1;
	  case 0:
		break;
	  default:
		_exit(0);
	}

	if (setsid() == -1) return -1;
	if (!nochdir) chdir("/");
	if (noclose) return 0;

	fd = open(_PATH_DEVNULL, O_RDWR, 0);
	if (fd != -1) {
		dup2(fd, STDIN_FILENO);
		dup2(fd, STDOUT_FILENO);
		dup2(fd, STDERR_FILENO);
		if (fd > 2) close(fd);
	}
	return 0;
}
